NetKet/Machine/fullconn_layer.hpp

```diff
@@ -0,0 +1,246 @@
+// Copyright 2018 The Simons Foundation, Inc. - All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef NETKET_FULLCONNLAYER_HH
+#define NETKET_FULLCONNLAYER_HH
+
+#include <Eigen/Dense>
+#include <complex>
+#include <fstream>
+#include <random>
+#include <vector>
+#include "Lookup/lookup.hpp"
+#include "Utils/all_utils.hpp"
+#include "abstract_layer.hpp"
+
+namespace netket {
+
+template <typename Activation, typename T>
+class FullyConnected : public AbstractLayer<T> {
+  using VectorType = typename AbstractLayer<T>::VectorType;
+  using MatrixType = typename AbstractLayer<T>::MatrixType;
+
+  Activation activation_;  // activation function class
+
+  bool usebias_;
+
+  int in_size_;        // input size
+  int out_size_;       // output size
+  int npar_;           // number of parameters in layer
+  MatrixType weight_;  // Weight parameters, W(in_size x out_size)
+  VectorType bias_;    // Bias parameters, b(out_size x 1)
+  MatrixType dw_;      // Derivative of weights
+  VectorType db_;      // Derivative of bias
+  VectorType z_;       // Linear term, z = W' * in + b
+  VectorType a_;       // Output of this layer, a = act(z)
+  VectorType din_;     // Derivative of the input of this layer.
+                       // Note that input of this layer is also the output of
+                       // previous layer
+
+  int mynode_;
+
+ public:
+  using StateType = typename AbstractLayer<T>::StateType;
+  using LookupType = typename AbstractLayer<T>::LookupType;
+
+  /// Constructor
+  FullyConnected(const int input_size, const int output_size,
+                 const bool use_bias = true)
+      : activation_(),
```

Comment:
Why do you call the constructor here?

Reply:
That is the activation class.

Reply:
Because not all layers need an activation. Maybe you have some pooling layers?

Reply:
Writing `activation_()` with an empty parameter list [will lead to the member being default-constructed](https://stackoverflow.com/questions/11164394/constructor-initialization-list-with-empty-initialization), which in this case [is equivalent](https://en.cppreference.com/w/cpp/language/data_members#Member_initialization) to not including `activation_` in the initializer list.

Whether to include it anyway is a matter of style, I guess.